diff --git a/themes/luci-theme-argon/ucode/template/themes/argon/header.ut b/themes/luci-theme-argon/ucode/template/themes/argon/header.ut
index 0637eb0..9d6fe43 100644
--- a/themes/luci-theme-argon/ucode/template/themes/argon/header.ut
+++ b/themes/luci-theme-argon/ucode/template/themes/argon/header.ut
@@ -115,6 +115,7 @@
 	{% if (node?.css): %}
 	<link rel="stylesheet" href="{{ resource }}/{{ node.css}}">
 	{% endif -%}
+	<link rel="stylesheet" href="{{ resource }}/zbt-mega-mobile.css?v={{ pkgs_update_time }}">
 	{% if (css): %}
 	<style title="text/css">
 		{{ css }}
@@ -122,3 +123,3 @@
-	<script src="{{ dispatcher.build_url('admin/translations', dispatcher.lang) }}?v={{ version.luciversion }}"></script>
-	<script src="{{ resource }}/cbi.js?v={{ version.luciversion }}"></script>
-	<script src="{{ resource }}/luci.js?v={{ version.luciversion }}"></script>
+	<script src="{{ dispatcher.build_url('admin/translations', dispatcher.lang) }}?v={{ pkgs_update_time }}"></script>
+	<script src="{{ resource }}/cbi.js?v={{ pkgs_update_time }}"></script>
+	<script src="{{ resource }}/luci.js?v={{ pkgs_update_time }}"></script>
